core/trino-main/src/main/java/io/trino/operator/aggregation/DistinctWindowAccumulator.java

```diff
@@ -0,0 +1,196 @@
+                + hash.getEstimatedSize();
+    }
+
+    @Override
+    public WindowAccumulator copy()
+    {
+        Page page = pageBuilder.build();
+        indexCurrentPage(page);
+        pageBuilder.reset();
+        return new DistinctWindowAccumulator(
+                delegate.copy(),
+                argumentTypes,
+                argumentChannels,
+                hash.copy(),
+                pagesIndexFactory);
+    }
+
+    @Override
+    public void addInput(WindowIndex index, int startPosition, int endPosition)
+    {
+        // index is remapped so just go from 0 to argumentChannels.size()
+        for (int position = startPosition; position <= endPosition; position++) {
+            if (pageBuilder.isFull()) {
+                Page page = pageBuilder.build();
+                indexCurrentPage(page);
+                pageBuilder.reset();
+            }
+            for (int channel = 0; channel < argumentChannels.size(); channel++) {
+                ValueBlock value = index.getSingleValueBlock(channel, position).getSingleValueBlock(0);
+                pageBuilder.getBlockBuilder(channel).append(value, 0);
+            }
+            pageBuilder.declarePosition();
+        }
+    }
+
+    private void indexCurrentPage(Page page)
+    {
+        Work<Block> work = hash.markDistinctRows(page);
+        checkState(work.process());
+        Block distinctMask = work.getResult();
+
+        int positionCount = distinctMask.getPositionCount();
+        checkArgument(positionCount == page.getPositionCount(), "Page position count dies not match distinct mask position count");
+        PagesIndex pagesIndex = pagesIndexFactory.newPagesIndex(argumentTypes, positionCount);
+
+        if (distinctMask instanceof RunLengthEncodedBlock) {
+            if (test(distinctMask, 0)) {
+                // all positions selected
+                pagesIndex.addPage(page);
+            }
+            // no positions selected
+            return;
+        }
+        else {
+            PageBuilder filteredPageBuilder = new PageBuilder(argumentTypes);
+            for (int position = 0; position < positionCount; position++) {
+                if (!test(distinctMask, position)) {
+                    continue;
+                }
+                for (int channel = 0; channel < argumentChannels.size(); channel++) {
+                    BlockBuilder filteredBlockBuilder = filteredPageBuilder.getBlockBuilder(channel);
+                    Block block = page.getBlock(channel);
+                    if (block instanceof ValueBlock valueBlock) {
+                        filteredBlockBuilder.append(valueBlock, position);
+                    }
+                    else {
+                        block.getUnderlyingValueBlock();
+                        ValueBlock singleValueBlock = block.getSingleValueBlock(position);
+                        filteredBlockBuilder.append(singleValueBlock, 0);
+                    }
```

Comment:
For appending use the method `Type.appendTo(Block block, int position, BlockBuilder blockBuilder);` instead of reasoning about Block types.

Reply:
nice :P